ring-container/src/sixsq/slipstream/server/ring_container.clj

```diff
@@ -0,0 +1,177 @@
+   optional cause."
+  ([msg]
+   (log-and-throw msg nil))
+  ([msg e]
+   (log/fatal msg)
+   (if e
+     (throw (ex-info (str msg "\n" e) {}))
+     (throw (ex-info msg {})))))
+
+(defn- as-symbol
+  "Converts argument to symbol or throws an exception."
+  [s]
+  (try
+    (symbol s)
+    (catch Exception e
+      (log-and-throw (str "invalid symbol: " s) e))))
+
+(defn- ns-and-var
+  "Provides the namespace and name of the symbol as symbols. If either is nil,
+   then an exception is thrown."
+  [s]
+  (let [result ((juxt namespace name) s)]
+    (if (some nil? result)
+      (log-and-throw (str "symbol (" s ") must be a complete, namespaced value"))
+      (map symbol result))))
+
+(defn- resolve-var
+  "Resolves the var in the given namespace or throws a descriptive exception."
+  [ns f]
+  (try
+    (ns-resolve (find-ns ns) f)
+    (catch Exception e
+      (log-and-throw (str "could not resolve " f " in namespace " ns) e))))
+
+(defn- dyn-resolve
+  "Dynamically requires the namespace of the given symbol and then resolves
+   the name of the symbol in this namespace. Returns the var for the resolved
+   symbol. Throws an exception if the symbol could not be resolved."
+  [s]
+  (let [[ns f] (ns-and-var (as-symbol s))]
+    (try
+      (require ns)
+      (catch Exception e
+        (log-and-throw (str "error requiring namespace: " ns))))
+
+    (if-let [result (resolve-var ns f)]
+      result
+      (log-and-throw (str "symbol (" s ") was not found")))))
+
+(defn- hook
+  "Creates a JVM shutdown hook that runs the given stopping function. Note
+   that some logging may be lost depending on the state of the JVM when the
+   hook is run."
+  [shutdown-fn]
+  (proxy [Thread] []
+    (run [] (shutdown-fn))))
+
+(defn- register-shutdown-hook
+  "Registers a shutdown hook in the JVM to shutdown the application and
+   application container cleanly."
+  [shutdown-fn]
+  (.. (Runtime/getRuntime)
+      (addShutdownHook (hook shutdown-fn))))
+
+(defn- create-shutdown-fn
+  [^Closeable server finalization-fn]
+  (fn [] (try
+           (and finalization-fn (finalization-fn))
+           (log/info "successfully finalized ring container")
+           (catch Exception e
+             (log/error "failure when finalizing ring container:" (str e)))
+           (finally
+             (try
+               (and server (.close server))
+               (log/info "successfully shutdown ring container")
+               (catch Exception e
+                 (log/error "failure when shutting down ring container:" (str e))))))))
+
+(defn- start-container
+  "Starts the aleph container with the given ring handler and on the given
+   port. Returns the server object that has been started, which can be stopped
+   by calling 'close' on the object."
+  [handler ^long port]
+  (log/info "starting aleph application container on port" port)
+  (let [start-server (dyn-resolve 'aleph.http/start-server)]
```

Comment:
When started as a daemon through the `main` function, the java class loader will be used.  As this class loader knows nothing about clojure source files, everything it needs to load the class must be available as compiled class files.  If aleph, environ, etc. were statically required, then the AOT compilation would also need to include those libraries as well, bloating the size of the jar file and causing potential conflicts with those dependencies on the classpath.

The AOT compilation of the`ring-container` namespace could be further reduced by removing the logging or also requiring it dynamically.  IMO, that seemed to be too much work for too little gain, but other opinions are welcome.